src/discourse.py

```diff
@@ -0,0 +1,288 @@
+        return _DiscourseTopicInfo(slug=path_components[-2], id_=path_components[-1])
+
+    def _retrieve_topic_first_post(self, url: str) -> dict:
+        """Retrieve the first post from a topic based on the URL to the topic.
+
+        Raises DiscourseError is pydiscourse raises an error or if the topic has been deleted.
+
+        Args:
+            usl: The URL to the topic.
+
+        Returns:
+            The first post from the topic.
+
+        """
+        topic_info = self._retrieve_topic_info_from_url(url=url)
+        try:
+            topic = self.client.topic(slug=topic_info.slug, topic_id=topic_info.id_)
+        except pydiscourse.exceptions.DiscourseError as discourse_error:
+            raise DiscourseError from discourse_error
+
+        first_post = next(
+            filter(lambda post: post["post_number"] == 1, topic["post_stream"]["posts"])
+        )
+        # Check for deleted topic
+        if first_post["user_deleted"]:
+            raise DiscourseError(f"topic has been deleted, {url=}")
+        return first_post
+
+    def check_topic_write_permission(self, url: str) -> bool:
+        """Check whether the credentials have write permission on a topic.
+
+        Raises DiscourseError if authentication fails or if the topic is not found.
+
+        Args:
+            url: The URL to the topic. Assume it includes the slug and id of the topic as the last
+                2 elements of the url.
+
+        Returns:
+            Whether the credentials have write permissions to the topic.
+
+        """
+        first_post = self._retrieve_topic_first_post(url=url)
+        return first_post["can_edit"]
+
+    def check_topic_read_permission(self, url: str) -> bool:
+        """Check whether the credentials have read permission on a topic.
+
+        Uses whether retrieve topic succeeds as indication whether the read permission is
+        available.
+
+        Raises DiscourseError if authentication fails or if the topic is not found.
+
+        Args:
+            url: The URL to the topic. Assume it includes the slug and id of the topic as the last
+                2 elements of the url.
+
+        Returns:
+            Whether the credentials have read permissions to the topic.
+
+        """
+        self._retrieve_topic_first_post(url=url)
+        return True
+
+    def retrieve_topic(self, url: str) -> str:
+        """Retrieve the topic content.
+
+        Raises DiscourseError if authentication fails, if the server refuses to return the
+        requested topic or if the topic is not found.
+
+        Args:
+            url: The URL to the topic. Assume it includes the slug and id of the topic as the last
+                2 elements of the url.
+
+        Returns:
+            The content of the first post in the topic.
+
+        """
+        first_post = self._retrieve_topic_first_post(url=url)
+        return first_post["cooked"].removeprefix("<p>").removesuffix("</p>")
```

Comment:
You are right, the stripping of the paragraph element should be removed on the next commit. This will be revisited when looking at how this works for markdown posted on discourse, it potentially means that documentation will always be updated rather than only on change which should be fine